elements/lisk-chain/src/data_access/storage.ts

```diff
@@ -12,291 +12,537 @@
-		);
+	public async getTempBlocks(): Promise<BlockJSON[]> {
+		const stream = this._db.createReadStream({
+			gte: getFirstPrefix(DB_KEY_TEMPBLOCKS_HEIGHT),
+			lte: getLastPrefix(DB_KEY_TEMPBLOCKS_HEIGHT),
+			reverse: true,
+		});
+		const tempBlocks = await new Promise<BlockJSON[]>((resolve, reject) => {
+			const blocks: BlockJSON[] = [];
+			stream
+				.on('data', ({ value }) => {
+					blocks.push(value);
+				})
+				.on('error', error => {
+					reject(error);
+				})
+				.on('end', () => {
+					resolve(blocks);
+				});
+		});
 
 		return tempBlocks;
 	}
 
 	public async isTempBlockEmpty(): Promise<boolean> {
-		const isEmpty = await this._storage.entities.TempBlock.isEmpty();
+		const stream = this._db.createReadStream({
+			gte: getFirstPrefix(DB_KEY_TEMPBLOCKS_HEIGHT),
+			lte: getLastPrefix(DB_KEY_TEMPBLOCKS_HEIGHT),
+			limit: 1,
+		});
+		const tempBlocks = await new Promise<BlockJSON[]>((resolve, reject) => {
+			const blocks: BlockJSON[] = [];
+			stream
+				.on('data', ({ value }) => {
+					blocks.push(value);
+				})
+				.on('error', error => {
+					reject(error);
+				})
+				.on('end', () => {
+					resolve(blocks);
+				});
+		});
 
-		return isEmpty;
+		return tempBlocks.length === 0;
 	}
 
 	public async clearTempBlocks(): Promise<void> {
-		await this._storage.entities.TempBlock.truncate();
+		await this._db.clear({
+			gte: getFirstPrefix(DB_KEY_TEMPBLOCKS_HEIGHT),
+			lte: getLastPrefix(DB_KEY_TEMPBLOCKS_HEIGHT),
+		});
 	}
 
 	public async deleteBlocksWithHeightGreaterThan(
 		height: number,
 	): Promise<void> {
-		await this._storage.entities.Block.delete({
-			// eslint-disable-next-line camelcase
-			height_gt: height,
-		});
+		const lastBlockHeader = await this.getLastBlockHeader();
+		const batchSize = 5000;
+		const loops = Math.ceil((lastBlockHeader.height - height + 1) / batchSize);
+		const start = height + 1;
+		// tslint:disable-next-line no-let
+		for (let i = 0; i < loops; i += 1) {
+			// Get all the required info
+			const startHeight = i * batchSize + start + i;
+			const endHeight = startHeight + batchSize - 1;
+			const headers = await this.getBlockHeadersByHeightBetween(
+				startHeight,
+				endHeight,
+			);
+			const blockIDs = headers.map(header => header.id);
+			const transactionIDs = [];
+			const batch = this._db.batch();
+			for (const blockID of blockIDs) {
+				try {
+					const ids = await this._db.get<string[]>(
+						`${DB_KEY_TRANSACTIONS_BLOCK_ID}${blockID}`,
+					);
+					transactionIDs.push(...ids);
+				} catch (error) {
+					if (!(error instanceof NotFoundError)) {
+						throw error;
+					}
+				}
+				batch.del(`${DB_KEY_BLOCKS_ID}${blockID}`);
+				batch.del(`${DB_KEY_TRANSACTIONS_BLOCK_ID}${blockID}`);
+			}
+			// tslint:disable-next-line no-let
+			for (let j = startHeight; j <= endHeight; j += 1) {
+				batch.del(`${DB_KEY_BLOCKS_HEIGHT}${formatInt(j)}`);
+			}
+			for (const txID of transactionIDs) {
+				batch.del(`${DB_KEY_TRANSACTIONS_ID}${txID}`);
+			}
```

Comment:
This will also be nicer as helper functions

Reply:
same as above